### PR TITLE
Fix `log_quasitriu` for internal scaling `s=0`

### LINKED ISSUE #1078: Matrix `exp` and `log` are not inverse of each other
1. Output of `versioninfo()`
 > Julia Version 1.10.3
Commit 0b4590a5507 (2024-04-30 10:59 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin22.4.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, apple-m1)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)
Environment:
  JULIA_EDITOR = emacsclient
  JULIA_INPUT_COLOR = bold
  JULIA_ANSWER_COLOR = bold

2. How Julia was installed

With [`juliaup`](https://github.com/JuliaLang/juliaup)

3. Minimal working example:

```julia
M = [0.9949357359852791 -0.015567763143324862 -0.09091193493947397 -0.03994428739762443 0.07338356301650806; 0.011813655598647289 0.9968988574699793 -0.06204555000202496 0.04694097614450692 0.09028834462782365; 0.092737943594701 0.059546719185135925 0.9935850721633324 0.025348893985651405 -0.018530261590167685; 0.0369187299165628 -0.04903571106913449 -0.025962938675946543 0.9977767446862031 0.12901494726320517; 0.0 0.0 0.0 0.0 1.0]
sdiff = exp(log(M)) - M
-log10(maximum(abs.(sdiff))) # < 3
```

The same test in Python gives 
```python
import numpy as np
import scipy as sp


M = np.array([[0.9949357359852791,-0.015567763143324862,-0.09091193493947397,-0.03994428739762443,0.07338356301650806],[0.011813655598647289,0.9968988574699793,-0.06204555000202496,0.04694097614450692,0.09028834462782365],[0.092737943594701,0.059546719185135925,0.9935850721633324,0.025348893985651405,-0.018530261590167685],[0.0369187299165628,-0.04903571106913449,-0.025962938675946543,0.9977767446862031,0.12901494726320517],[0.0,0.0,0.0,0.0,1.0]])

sdiff = sp.linalg.expm(sp.linalg.logm(M)) - M

print(-np.log10(np.max(np.abs(sdiff)))) # > 15
```

### PR DESCRIPTION
**Pull request by aravindh-krishnamoorthy**
This PR is a potential fix for JuliaLang/LinearAlgebra.jl#1078.

## Description
The function https://github.com/JuliaLang/julia/blob/2a06376c18afd7ec875335070743dcebcd85dee7/stdlib/LinearAlgebra/src/triangular.jl#L2220 computes $\boldsymbol{A}^{\dfrac{1}{2^s}} - \boldsymbol{I}$ for a real-valued $2\times 2$ matrix $\boldsymbol{A}$ using Algorithm 5.1 in [R1]. However, the algorithm in [R1] as well as the above function do not handle the case $s=0.$ This fix extends the function to compute $\boldsymbol{A}^{\dfrac{1}{2^s}} - \boldsymbol{I} \Bigg|_{s=0} = \boldsymbol{A} - \boldsymbol{I}.$

## Checklist
- [X] Fix code: `stdlib\LinearAlgebra\src\triangular.jl` in function `_sqrt_pow_diag_block_2x2!(A, A0, s)`.
- [X] Add test case: `stdlib\LinearAlgebra\test\triangular.jl`.
- [X] Update `NEWS.md`.
- [X] Testing and self review.

|  Tag  | Reference |
| --- | --- |
| <nobr>[R1]</nobr> | Al-Mohy, Awad H. and Higham, Nicholas J. "Improved Inverse Scaling and Squaring Algorithms for the Matrix Logarithm", 2011, url: https://eprints.maths.manchester.ac.uk/1687/1/paper11.pdf |

### CODE REVIEW

**Comment by dkarrasch** [comment]
I checked locally and this also solves the example in https://github.com/JuliaLang/julia/issues/54833#issuecomment-2180915810.

**Comment by dkarrasch** [comment]
Do you see if https://github.com/scipy/scipy/blob/628a6b3bf3e1a3149b8a6390322ae976bba930fa/scipy/linalg/_matfuncs_inv_ssq.py#L367-L377 and subsequent code handles it the same way? IIRC then the scipy parameters for the failing example included `s=0` as well.

File: NEWS.md [hunk 1]
```diff
@@ -148,6 +148,7 @@ Standard library changes
   Custom array types may specialize this function to return an appropriate result ([#55252]).
 * The matrix multiplication `A * B` calls `matprod_dest(A, B, T::Type)` to generate the destination.
   This function is now public ([#55537]).
+* Now `exp(log(A))` for matrix `A` correctly returns `A`, up to numerical precision ([#56311]).
```
Comment (oscardssmith) [inline NEWS.md:151]: the issue here is purely in `log`, right?

Comment (oscardssmith) [reply NEWS.md:151]: If so, possibly rephrase to something like "fixed numerical stability of `Matrix` logarithm in some cases".

Comment (aravindh-krishnamoorthy) [inline NEWS.md:151]:
```suggestion
* Fixed numerical stability (in some cases) of `log(A)` for matrix `A` ([#56311]).
```

**Comment by dkarrasch** [comment]
Honestly, I don't think we should mention this in NEWS. We should instead backport this to v1.10 and v1.11.

File: stdlib/LinearAlgebra/src/triangular.jl [hunk 2]
```diff
@@ -2218,6 +2218,11 @@ end
 # SIAM J. Sci. Comput., 34(4), (2012) C153–C169. doi: 10.1137/110852553
 # Algorithm 5.1
 Base.@propagate_inbounds function _sqrt_pow_diag_block_2x2!(A, A0, s)
+    if iszero(s)
+        A[1,1] -= 1
+        A[2,2] -= 1
+        return
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/triangular.jl:2224]:
```suggestion
        return A
```

to be consistent and type-stable with the following "branch".

**Comment by aravindh-krishnamoorthy** [comment]
> Do you see if https://github.com/scipy/scipy/blob/628a6b3bf3e1a3149b8a6390322ae976bba930fa/scipy/linalg/_matfuncs_inv_ssq.py#L367-L377 and subsequent code handles it the same way? IIRC then the scipy parameters for the failing example included `s=0` as well.

Thank you for this reference. In Julia function `_log_quasitriu!`: https://github.com/JuliaLang/julia/blob/2a06376c18afd7ec875335070743dcebcd85dee7/stdlib/LinearAlgebra/src/triangular.jl#L2014 **[R1, Algorithm 4.1]** is used just to compute parameters `m` and `s`. Then, $\boldsymbol{A}^{\dfrac{1}{2^s}} - \boldsymbol{I}$ is computed using **[R1, Algorithm 5.1]**.

On the other hand, in the python code in https://github.com/scipy/scipy/blob/628a6b3bf3e1a3149b8a6390322ae976bba930fa/scipy/linalg/_matfuncs_inv_ssq.py#L367
a square root operation is performed every time `s` is incremented. Hence, at the end of the algorithm in
https://github.com/scipy/scipy/blob/628a6b3bf3e1a3149b8a6390322ae976bba930fa/scipy/linalg/_matfuncs_inv_ssq.py#L422 
$\boldsymbol{A}^{\left(\frac{1}{2}\right)^s} - \boldsymbol{I}$ is computed. So, it does not inherit the problem in **[R1, Algorithm 5.1]**.

**Review by dkarrasch — APPROVED** [review]

**Comment by jishnub** [comment]
Good to merge?

**Comment by oscardssmith** [comment]
thanks for this!

**Pull request merged**